File: app/src/test/java/com/tjb/dwf/MainActivityTest.java
package com.tjb.dwf;

import android.app.Instrumentation;
import android.content.Intent;
import android.os.Build;
import android.util.Log;

import androidx.test.espresso.intent.Intents;
import androidx.test.rule.ActivityTestRule;

import org.hamcrest.Matcher;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static androidx.test.espresso.intent.Intents.intended;
import static androidx.test.espresso.intent.Intents.intending;
import static androidx.test.espresso.intent.Intents.times;
import static androidx.test.espresso.intent.matcher.IntentMatchers.anyIntent;
import static androidx.test.espresso.intent.matcher.IntentMatchers.hasComponent;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = Build.VERSION_CODES.P)
public class MainActivityTest {
    @Rule
    public ActivityTestRule<MainActivity> mainActivityTestRule =
            new ActivityTestRule<>(MainActivity.class, true, false);

    @Test
    public void whenNoUserInIntent_mainActivity_launchesLoginActivity() {
        Intents.init();

        // Stub the login intent to prevent LoginActivity from being displayed.
        // This helps to fix the Espresso timeout exception.
        Matcher<Intent> loginIntent = hasComponent(LoginActivity.class.getName());
        intending(loginIntent).respondWith(new Instrumentation.ActivityResult(0, null));

        mainActivityTestRule.launchActivity(null);
        intended(hasComponent(LoginActivity.class.getName()));

        Intents.release();
    }

    @Test
    public void whenUserInIntent_mainActivity_staysOpen() {
        Intents.init();

        Intent startIntent = new Intent();
        UserPojo dummyUser = new UserPojo("f", "l", "t", 0);
        startIntent.putExtra(UserPojo.SERIALIZE_TAG, dummyUser);
        mainActivityTestRule.launchActivity(startIntent);

        // how to say, "intended matches any 0 times
        intended(anyIntent(), times(0));

        Intents.release();
    }
}

File: app/src/test/java/com/tjb/dwf/MainActivityTest.kt
package com.tjb.dwf;

import android.os.Build
import android.view.View
import androidx.test.rule.ActivityTestRule
import com.tjb.dwf.di.MockPinchGestureListener
import com.tjb.dwf.di.MockUserModule
import com.tjb.dwf.main.MainActivity
import com.tjb.dwf.user.UserPojo
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.P])
class MainActivityTest {
    @get:Rule
    var mainActivityTestRule = ActivityTestRule(MainActivity::class.java, true, false)

    @Before
    fun setup() {
        every {
            MockPinchGestureListener.getMockPinchGestureListener().installMainActivity(any())
        } returns Unit
        val user = createUserPojo()
        every {
            MockUserModule.getMockUserController().getUser()
        } returns user
        mainActivityTestRule.launchActivity(null)
    }

    @Test
    fun whenOnClickLogout_userControllerLogout_isCalled() {
        every {
            MockUserModule.getMockUserController().logout(any())
        } returns Unit
        val v = mockk<View>(relaxed = true)
        mainActivityTestRule.activity.onClickLogout(v)

        verify(exactly = 1) {
            MockUserModule.getMockUserController().logout(mainActivityTestRule.activity)
        }
    }

    @Test
    fun onCreate_pinchGestureListener_installCalled() {
        verify(exactly = 1) {
            MockPinchGestureListener.getMockPinchGestureListener().installMainActivity(mainActivityTestRule.activity)
        }
    }

    private fun createUserPojo(): UserPojo {
        return UserPojo("u", "t", 0)
    }
}
